staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcdfeature/feature_support_checker.go

```diff
@@ -0,0 +1,183 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdfeature
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+var (
+	// This is a error type that we used in checking RequestWatchProgress support
+	errNoSupportRequestWatchProgress = errors.New("requestWatchProgress is not supported for this cluster")
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+
+	// DefaultFeatureSupportChecker is a shared global FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) bool
+	// This method works with etcd client to recalcuate feature support and caches it.
+	EtcdClientSupports(ctx context.Context, c client, feature storage.Feature) (bool, error)
+	// Only use for testing and use defer afterwork.
+	SetFeatureSupportCheckerDuringTest(feature storage.Feature, value bool) func()
+}
+
+// featureSupportInfo holds information about feature support status, used for cache.
+type featureSupportInfo struct {
+	Supported bool
+}
+
+type defaultFeatureSupportChecker struct {
+	// featureSupportCache store result of if feature is supported or not.
+	featureSupportCache sync.Map
+}
+
+func newDefaultFeatureSupportChecker() *defaultFeatureSupportChecker {
+	return &defaultFeatureSupportChecker{}
+}
+
+// Supports can check the featue from anywhere without storage if it was checked before with Supports
+// Only return true or false if it was defined or not.
+func (f *defaultFeatureSupportChecker) Supports(feature storage.Feature) bool {
+	if info, ok := f.featureSupportCache.Load(feature); ok {
+		if val, ok := info.(featureSupportInfo); ok {
+			return val.Supported
+		}
+	}
+	return false
+}
+
+// EtcdClientSupports checks whether a specific feature is supported by etcd and caches it.
+// TODO: Extend this implementation to support feature-checking for the cacher.
+func (f *defaultFeatureSupportChecker) EtcdClientSupports(ctx context.Context, c client, feature storage.Feature) (bool, error) {
+	// Check if the enabled feature is already cached.
+	if info, ok := f.featureSupportCache.Load(feature); ok {
+		if val, ok := info.(featureSupportInfo); ok {
+			return val.Supported, nil
```

Comment:
This is not a correct. It will cache results from the first etcd cluster, while apiserver might connect to different apiservers. Please don't cache per feature, we don't support or care about other features. Please cache results per endpoint.

Reply:
Updated! 
PTAL. Thanks.